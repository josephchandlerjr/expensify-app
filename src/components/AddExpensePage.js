import React from 'react'
import ReactDOM from 'react-dom'

const AddExpensePage = () => {
    return (
        <div>This is from my add expense component</div>
    )
}

export default AddExpensePage