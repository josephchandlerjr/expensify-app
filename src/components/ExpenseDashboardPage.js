import React from 'react'
import ReactDOM from 'react-dom'

const ExpenseDashboardPage = () => {
    return (
        <div>This is from my dashboard component</div>
    )
}

export default ExpenseDashboardPage