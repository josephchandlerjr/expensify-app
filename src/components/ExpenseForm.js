import React from 'react'
import moment from 'moment'
import { SingleDatePicker } from 'react-dates'


export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props)    
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: ''
        }
        this.onDescriptionChange = this.onDescriptionChange.bind(this)
        this.onNoteChange = this.onNoteChange.bind(this)
        this.onAmountChange = this.onAmountChange.bind(this)
        this.onDateChange = this.onDateChange.bind(this)
        this.onFocusChange = this.onFocusChange.bind(this)
        this.onSubmit = this.onSubmit.bind(this)
    }
    onDescriptionChange(e) {
        const description = e.target.value
        this.setState( () => ({
            description
        }))
    }
    onNoteChange(e) {
        const note = e.target.value
        this.setState( () => ({
            note
        }))
    }
    onAmountChange(e) {
        const amount = e.target.value
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState( () => ({
                amount
            })) 
        }
    }
    onDateChange(createdAt) {
        if(!createdAt) return
        this.setState( () => ({ createdAt }))
    }
    onFocusChange({ focused }) {
        this.setState( () => ({
            calendarFocused: focused
        }))
    }
    onSubmit(e) {
        e.preventDefault()
        if (!this.state.description || !this.state.amount) {
            this.setState( () => ({ error: 'Please provide a description and amount.' })) 
        } else {
            this.setState( () => ({ error: '' })) 
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100, //working in cents
                createdAt: this.state.createdAt.valueOf(), // unix timestamp in milliseconds
                note: this.state.note
            })
        }
    }
    render() {
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input type="text" 
                           placeholder="Description"
                           value={this.state.description}
                           autoFocus    
                           onChange={this.onDescriptionChange}
                           />
                    <input type="text"
                           placeholder="Amount"
                           value={this.state.amount}
                           onChange={this.onAmountChange}
                           />
                    <textarea 
                        value={this.state.note}
                        onChange={this.onNoteChange}
                        placeholder="Add a note for your expense."
                        >
                        </textarea>
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={ (day) => false}
                        />
                    <button>Add Expense</button>
                </form>
            </div>
        )
        
    }
}