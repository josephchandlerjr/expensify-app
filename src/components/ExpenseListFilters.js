import React from 'react'
import { connect } from 'react-redux'
import { sortByAmount, sortByDate } from '../actions/filters'
import { setTextFilter, setStartDate, setEndDate } from '../actions/filters'
import { DateRangePicker } from 'react-dates'

export class ExpenseListFilters extends React.Component {
    constructor(props) {
        super(props)
        this.state = {
            calendarFocused: null
        }

        this.onDatesChange = this.onDatesChange.bind(this)
        this.onFocusChange = this.onFocusChange.bind(this)
        this.onSortChange = this.onSortChange.bind(this)
        this.onTextChange = this.onTextChange.bind(this)
    }
    onDatesChange({ startDate, endDate}) {
        this.props.setStartDate(startDate)
        this.props.setEndDate(endDate)
    }
    onFocusChange(calendarFocused) {
        this.setState({ calendarFocused })
    }
    onTextChange(evt) {
        this.props.setTextFilter(evt.target.value)
    }
    onSortChange(evt) {
        const sortBy = evt.target.value
        if (sortBy === 'amount') this.props.sortByAmount()
        if (sortBy === 'date') this.props.sortByDate()
    }
    render() {
        return (
            <div>
                <input type="text" value={this.props.filters.text} onChange={this.onTextChange}
                />
                <select value={this.props.filters.sortBy}
                    onChange={this.onSortChange}>
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker
                    startDate={this.props.filters.startDate}
                    startDateId={";oiahsd;oifhjao;ishjfo;adishjfdois;f"} 
                    endDate={this.props.filters.endDate}
                    endDateId={"ewoiueriuyeroeyroei"}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    isOutsideRange={ (day) => false}
                    numberOfMonths={1}
                    />
            </div>
        )
    }
}
    

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    }
} 

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setStartDate: (date) => dispatch(setStartDate(date)),
    setEndDate: (date) => dispatch(setEndDate(date))
})
export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters)