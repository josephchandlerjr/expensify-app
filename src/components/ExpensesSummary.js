import React from 'react'
import { connect } from 'react-redux'
import numeral from 'numeral'
import selectExpenses from '../selectors/expenses'
import expensesTotal from '../selectors/expenses-total'

export const ExpensesSummary = (props) => (
    <div>
        <h4>
            {
                `
                Viewing ${props.expensesCount} 
                expense${props.expensesCount === 1 ? '': 's'}
                totaling ${numeral(props.expensesTotal / 100).format('$0,0.00')}
                `
            }
        </h4>
        <h3>{props.expenseTotal}</h3>
    </div>
)

const mapStateToProps = (state) => {
    const expenses = selectExpenses(state.expenses, state.filters)
    return {
        expensesCount: expenses.length,
        expensesTotal: expensesTotal(expenses)
        }
}

export default connect(mapStateToProps)(ExpensesSummary)