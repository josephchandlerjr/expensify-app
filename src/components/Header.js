import React from 'react'
import ReactDOM from 'react-dom'
import { NavLink } from 'react-router-dom'
const Header = () => {
    return (
        <header>
            <h1>Expensify</h1>
            <NavLink to="/" activeClassName="is-active" exact={true}>Dashboard</NavLink>
            <NavLink to="/create" activeClassName="is-active">Create</NavLink>
            <NavLink to="/edit" activeClassName="is-ac-aive">Edit</NavLink>
            <NavLink to="/help" activeClassName="is-active">Help</NavLink>
        </header>
    )
}

export default Header