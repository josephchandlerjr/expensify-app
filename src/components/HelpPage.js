import React from 'react'
import ReactDOM from 'react-dom'

const HelpPage = () => {
    return (
        <div>This is from my help page component</div>
    )
}

export default HelpPage