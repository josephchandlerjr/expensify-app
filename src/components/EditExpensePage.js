import React from 'react'
import ExpenseForm from './ExpenseForm'
import { connect } from 'react-redux'
import { editExpense } from '../actions/expenses'
import { removeExpense } from '../actions/expenses'

export class EditExpensePage extends React.Component {
    constructor(props) {
        super(props)
        this.onSubmit = this.onSubmit.bind(this)
        this.remove = this.remove.bind(this)
    }
    onSubmit(expense) {
        this.props.onSubmit(this.props.expense.id, expense)
        this.props.history.push('/')
    }
    remove() {
        this.props.remove(this.props.expense.id)
        this.props.history.push('/')
    }
    render() {
        return (
            <div>
                <ExpenseForm 
                onSubmit={ this.onSubmit }
                expense={this.props.expense} />
                <button onClick={ this.remove }>Remove</button>
            </div>
        )  
    }
}
 
const mapStateToProps = (state, props) => ({
    expense: state.expenses.find( (expense) => expense.id === props.match.params.id )
})

const mapDispatchToProps = (dispatch) => ({
    onSubmit: (id, expense) => dispatch(editExpense(id, expense)),
    remove: (id) => dispatch(removeExpense({ id }))
})

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage)