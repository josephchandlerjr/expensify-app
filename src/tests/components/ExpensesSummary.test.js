import React from 'react'
import { ExpensesSummary } from '../../components/ExpensesSummary'
import { shallow } from 'enzyme'
import expenses from '../fixtures/expenses'

let wrapper

beforeEach( () => {
    
})

test('should properly render expenses summary with no expenses', () => {
    wrapper = shallow(<ExpensesSummary  expensesCount={0}
                                        expensesTotal={0} />)
    expect(wrapper).toMatchSnapshot()
})

test('should properly render expenses summary', () => {
    wrapper = shallow(<ExpensesSummary  expensesCount={3}
                                        expensesTotal={500} />)
    expect(wrapper).toMatchSnapshot()
})
