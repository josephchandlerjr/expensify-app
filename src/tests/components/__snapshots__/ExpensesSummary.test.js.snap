// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should properly render expenses summary 1`] = `
<div>
  <h4>
    
                Viewing 3 
                expenses
                totaling $5.00
                
  </h4>
  <h3 />
</div>
`;

exports[`should properly render expenses summary with no expenses 1`] = `
<div>
  <h4>
    
                Viewing 0 
                expenses
                totaling $0.00
                
  </h4>
  <h3 />
</div>
`;
